### PR TITLE
En 5649 reject headers with nonce under final

### PR DESCRIPTION
**Pull request by iulianpascalau**
added check on the header interceptors not to process old, under finality nonce block headers. This will prevent spam.

### CODE REVIEW

File: cmd/node/factory/interface.go [hunk 1]
```diff
@@ -19,4 +20,8 @@ type HeaderSigVerifierHandler interface {
 	VerifyRandSeedAndLeaderSignature(header data.HeaderHandler) error
 	VerifySignature(header data.HeaderHandler) error
 	IsInterfaceNil() bool
-}
\ No newline at end of file
+}
+
+type interceptedDataFactoryWithFinalAttesterSetter interface {
```
Comment (sasurobert) [inline cmd/node/factory/interface.go]: add comment to interface.

File: cmd/node/factory/structs.go [hunk 2]
```diff
@@ -780,6 +785,26 @@ func ProcessComponentsFactory(args *processComponentsFactoryArgs) (*Process, err
...
+func applyblockTrackerOnInterceptors(
+	finalityAttester process.FinalityAttester,
+	interceptorsContainer process.InterceptorsContainer,
+) error {
+
+	var foundErr error
+	interceptorsContainer.Iterate(func(key string, interceptor process.Interceptor) bool {
+		dataFactory := interceptor.InterceptedDataFactory()
+		setter, ok := dataFactory.(interceptedDataFactoryWithFinalAttesterSetter)
+		if ok {
```
Comment (sasurobert) [inline cmd/node/factory/structs.go]: if !ok return true else setter.set and return foundErr == nil. If you have a set of errors, you only catch the last one. Should save founderr into an other variable.

File: integrationTests/testProcessorNode.go [hunk 3]
```diff
@@ -1320,3 +1321,16 @@ func (tpn *TestProcessorNode) initHeaderValidator() {
 
 	tpn.HeaderValidator, _ = block.NewHeaderValidator(argsHeaderValidator)
 }
+
+func (tpn *TestProcessorNode) applyBlockTrackerOnInterceptors() {
+	tpn.InterceptorsContainer.Iterate(func(key string, interceptor process.Interceptor) bool {
+		dataFactory := interceptor.InterceptedDataFactory()
+		setter, ok := dataFactory.(interceptedDataFactoryWithFinalAttesterSetter)
+		if ok {
+			foundErr := setter.SetFinalityAttester(tpn.BlockTracker)
```
Comment (sasurobert) [inline integrationTests/testProcessorNode.go]: same comments as in structs.go

File: process/interceptors/factory/nilFinalityAttester.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/logger"
+)
+
+var log = logger.GetOrCreate("process/interceptors/factory")
+
+// nilFinalityAttester is a nil implementation that will cause all received headers
```
Comment (sasurobert) [inline process/interceptors/factory/nilFinalityAttester.go]: that returns default value, without checking actual finality.

File: process/mock/finalityAttesterStub.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+type FinalityAttesterStub struct {
+	GetFinalHeaderCalled func(shardID uint32) (data.HeaderHandler, []byte, error)
+}
+
+func (fas *FinalityAttesterStub) GetFinalHeader(shardID uint32) (data.HeaderHandler, []byte, error) {
+	if fas.GetFinalHeaderCalled != nil {
+		return fas.GetFinalHeaderCalled(shardID)
+	}
+
+	return &block.Header{
```
Comment (sasurobert) [inline process/mock/finalityAttesterStub.go]: define header separately like in nilFinalityAttester

File: process/track/baseBlockTrack.go [hunk 6]
```diff
@@ -78,6 +92,20 @@ func (bbt *baseBlockTrack) receivedMetaBlock(headerHandler data.HeaderHandler, m
 		return
 	}
 
+	err := bbt.checkBlockBasicValidity(metaBlock)
```
Comment (sasurobert) [inline process/track/baseBlockTrack.go]: no need for this as implementation is nil. log.Trace will be never called, neither return.

File: process/track/baseBlockTrack.go [hunk 7]
```diff
@@ -89,6 +117,10 @@ func (bbt *baseBlockTrack) receivedMetaBlock(headerHandler data.HeaderHandler, m
 	bbt.blockProcessor.processReceivedHeader(metaBlock)
 }
 
+func (bbt *baseBlockTrack) checkBlockBasicValidity(_ data.HeaderHandler) error {
```
Comment (sasurobert) [inline process/track/baseBlockTrack.go]: rename checkBasicBlockValidity. However I do not think this is needed. Interceptors are doing the basic validity.

File: process/interface.go [hunk 8]
```diff
@@ -637,3 +640,9 @@ type EpochStartDataCreator interface {
 	VerifyEpochStartDataForMetablock(metaBlock *block.MetaBlock) error
 	IsInterfaceNil() bool
 }
+
+// FinalityAttester is able to manage the final blocks
+type FinalityAttester interface {
```
Comment (sasurobert) [inline process/interface.go]: Maybe rename FinalityAttester to something nice :).

File: process/interceptors/factory/interceptedMetaHeaderDataFactory_test.go [hunk 9]
```diff
@@ -141,3 +141,27 @@ func TestNewInterceptedMetaHeaderDataFactory_ShouldWorkAndCreate(t *testing.T) {
...
+	t.Parallel()
+
+	arg := createMockArgument()
+	imh, _ := NewInterceptedMetaHeaderDataFactory(arg)
+	newAttester := &nilFinalityAttester{}
+
+	err := imh.SetFinalityAttester(newAttester)
+
+	assert.Nil(t, err)
+	assert.True(t, imh.finalityAttester == newAttester)
```
Comment (AdoAdoAdo) [inline process/interceptors/factory/interceptedMetaHeaderDataFactory_test.go]: why not assert equal?

File: process/interceptors/factory/interceptedShardHeaderDataFactory_test.go [hunk 10]
```diff
@@ -80,3 +80,27 @@ func TestInterceptedShardHeaderDataFactory_ShouldWorkAndCreate(t *testing.T) {
...
+	t.Parallel()
+
+	arg := createMockArgument()
+	ish, _ := NewInterceptedShardHeaderDataFactory(arg)
+	newAttester := &nilFinalityAttester{}
+
+	err := ish.SetFinalityAttester(newAttester)
+
+	assert.Nil(t, err)
+	assert.True(t, ish.finalityAttester == newAttester)
```
Comment (AdoAdoAdo) [inline process/interceptors/factory/interceptedShardHeaderDataFactory_test.go]: why not assert.Equal

File: process/interceptors/factory/nilFinalityAttester.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+//not to be actually checked
+type nilFinalityAttester struct {
```
Comment (AdoAdoAdo) [inline process/interceptors/factory/nilFinalityAttester.go]: I don't like the this mock in production code, this is only used in tests right?

I would move this to the mock folder instead
You also don't need any unit tests over a mock implementation

File: process/interceptors/singleDataInterceptor_test.go [hunk 12]
```diff
@@ -82,14 +83,16 @@ func TestNewSingleDataInterceptor_NilInterceptorThrottlerShouldErr(t *testing.T)
...
+		factory,
 		&mock.InterceptorProcessorStub{},
 		&mock.InterceptorThrottlerStub{},
 	)
 
-	assert.NotNil(t, sdi)
-	assert.Nil(t, err)
+	require.False(t, check.IfNil(sdi))
+	require.Nil(t, err)
+	assert.True(t, sdi.InterceptedDataFactory() == factory)
```
Comment (AdoAdoAdo) [inline process/interceptors/singleDataInterceptor_test.go]: why not assert.equal?

File: process/track/baseBlockTrack.go (hunk 7 above)
Comment (AdoAdoAdo) [inline process/track/baseBlockTrack.go]: why is this needed? does it need to implement any interface that requires this? otherwise I would remove it.

File: cmd/node/factory/interface.go (hunk 1 above)
Comment (iulianpascalau) [reply cmd/node/factory/interface.go]: removed, no need for this at it will be injected

File: cmd/node/factory/structs.go (hunk 2 above)
Comment (iulianpascalau) [reply cmd/node/factory/structs.go]: function removed, no longer necessary

File: integrationTests/testProcessorNode.go (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/testProcessorNode.go]: function removed

File: process/interceptors/factory/nilFinalityAttester.go (hunk 4 above)
Comment (iulianpascalau) [reply process/interceptors/factory/nilFinalityAttester.go]: implementation removed, no longer necessary

File: process/interface.go (hunk 8 above)
Comment (iulianpascalau) [reply process/interface.go]: Do you have a suggestion? FinalityAttester describes best what this interface should suppose to do: attest finality (in case of block headers)

File: process/block/interceptedBlocks/common.go [hunk 13]
```diff
@@ -77,6 +82,24 @@ func checkHeaderHandler(hdr data.HeaderHandler) error {
...
 
+func checkHeaderHandlerFinality(hdr data.HeaderHandler, attester process.FinalityAttester) error {
+	attestedHdr, _, err := attester.GetFinalHeader(hdr.GetShardID())
+	if err != nil {
+		return err
+	}
+
+	isHeaderReceivedUnderFinal := hdr.GetNonce() < attestedHdr.GetNonce()
+	if isHeaderReceivedUnderFinal {
+		return fmt.Errorf("%w while checking attested hdr, got nonce: %d, final hdr nonce: %d",
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/block/interceptedBlocks/common.go]: ok, renamed to `checkHeaderHandlerHasLowerNonceThanFinalHeader`

File: process/interceptors/factory/interceptedMetaHeaderDataFactory_test.go (hunk 9 above)
Comment (iulianpascalau) [reply process/interceptors/factory/interceptedMetaHeaderDataFactory_test.go]: Test removed. The assert.True tested that the same pointer has been set (set worked correctly)

File: process/interceptors/factory/interceptedShardHeaderDataFactory_test.go (hunk 10 above)
Comment (iulianpascalau) [reply process/interceptors/factory/interceptedShardHeaderDataFactory_test.go]: Test removed. The assert.True tested that the same pointer has been set (set worked correctly)

File: process/interceptors/factory/nilFinalityAttester.go (hunk 11 above)
Comment (iulianpascalau) [reply process/interceptors/factory/nilFinalityAttester.go]: Nil implementation removed, no longer required as we inject the implementation on the constructor

File: process/interceptors/singleDataInterceptor_test.go (hunk 12 above)
Comment (iulianpascalau) [reply process/interceptors/singleDataInterceptor_test.go]: Test removed. The assert.True tested that the same pointer has been set (set worked correctly)

File: process/track/baseBlockTrack.go (hunk 7 above)
Comment (iulianpascalau) [reply process/track/baseBlockTrack.go]: function removed

Comment (iulianpascalau) [reply process/track/baseBlockTrack.go]: function removed

File: process/track/baseBlockTrack.go (hunk 6 above)
Comment (iulianpascalau) [reply process/track/baseBlockTrack.go]: function removed

File: process/mock/finalityAttesterStub.go (hunk 5 above)
Comment (iulianpascalau) [reply process/mock/finalityAttesterStub.go]: done

File: process/interface.go (hunk 8 above)
Comment (iulianpascalau) [reply process/interface.go]: HeaderFinalityAttester? @AdoAdoAdo

File: process/track/baseBlockTrack.go [hunk 14]
```diff
@@ -275,6 +276,64 @@ func (bbt *baseBlockTrack) GetCrossNotarizedHeader(shardID uint32, offset uint64
 	return bbt.crossNotarizer.getNotarizedHeader(shardID, offset)
 }
 
+// CheckBlockBasicValidity checks if the given header is valid related to some basic checks
+func (bbt *baseBlockTrack) CheckBlockBasicValidity(headerHandler data.HeaderHandler) error {
+	if check.IfNil(headerHandler) {
+		return fmt.Errorf("nil header handler")
+	}
+
+	finalHeader, _, err := bbt.GetFinalHeader(headerHandler.GetShardID())
```
Comment (AdoAdoAdo) [inline process/track/baseBlockTrack.go]: this error is misleading, this function should only check if the block is valid or invalid.
This error is unrelated to this specific block, so it shouldn't be here.

Comment (iulianpascalau) [reply process/track/baseBlockTrack.go]: done, split the function in 2
